underwater robot/uuv extension

DanielDuecker opened the pull request:
This PR corresponds to https://github.com/PX4/Firmware/pull/14079

It extends the underwater robot capabilities and provides an uuv_motor_model  allows to model the esc-deadzone which is important for vehicle actuators  which can spin in both directions.

The old hippocampus model is kept for the moment as part of the px4-firmware tutorial depends on it. However it likely makes sense to have the tutorial depending on the new model at some point.

PhilippHastedt committed:
Add UUV

timzarhansen committed:
update_own_repo

timzarhansen committed:
tmp

timzarhansen committed:
added yaw=0 current yaw

DanielDuecker committed:
integrate existing uuv features

Merge branch 'hippocampus' of https://github.com/zarbokk/sitl_gazebo into uuv_extension

DanielDuecker committed:
major renaming from uuv2 -> uuv_hippocampus

DanielDuecker committed:
minor bug fixes and clean up

DanielDuecker committed:
* added uuv_hippocampus model carrying a camera

* general code clean up + comments

DanielDuecker committed:
added uuv_hippocampus_cam.world

DanielDuecker committed:
minor changes to hippoc_cam world

DanielDuecker committed:
cleaning up

DanielDuecker committed:
minor changes to new uuv_hippocampus world

DanielDuecker committed:
Merge branch 'uuv_extension' of https://github.com/danielduecker/sitl_gazebo into uuv_extension

DanielDuecker committed:
Merge branch 'uuv_extension' of https://github.com/danielduecker/sitl_gazebo into uuv_extension

DanielDuecker committed:
Merge branch 'master' of https://github.com/px4/sitl_gazebo into uuv_extension

DanielDuecker committed:
remove  >> frame='' << from pose in sdf files

Jaeyoung-Lim commented on src/gazebo_uuv_plugin.cpp:
```diff
@@ -123,58 +135,51 @@ void GazeboUUVPlugin::OnUpdate(const common::UpdateInfo& _info) {
+      << command_[0] << ","
+      << command_[1] << ","
+      << command_[2] << ","
+      << command_[3] << "\n";
```
This will flood the console.

DanielDuecker replied:
Thanks, removed.

Jaeyoung-Lim commented on src/gazebo_uuv_plugin.cpp:
```diff
@@ -123,58 +135,51 @@ void GazeboUUVPlugin::OnUpdate(const common::UpdateInfo& _info) {
+      //rotor_links_[i+1]->AddRelativeTorque(rotor_torque);
+      torques[i] = rotor_torque[0];
+    }
+    //std::cout << std::showpos << std::fixed << std::setprecision(4) << "Force:" << forces[0] << "\n";
```
Is there a reason why all these comments are needed?

DanielDuecker replied:
No they should be gone now (debugging left-overs)

Jaeyoung-Lim commented on src/gazebo_uuv_plugin.cpp:
```diff
@@ -96,23 +89,42 @@ void GazeboUUVPlugin::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf) {
     << command_[2] << ","
-    << command_[3] << ","
-    << "\n"; */
+    << command_[3] << "\n";*/
```
If this is not used, please remove

DanielDuecker replied:
should be gone now

Jaeyoung-Lim reviewed (commented):
Thanks for the contribution! I have a few general comments as the following.

- The name of the plugin is `gazebo_uuv_motor_model`. However, what the plugin is actually doing is simulating the uuv dynamics. It includes the combined dynamics of 4 motors and the damping dynamics of the body. I think this is confusing, since the `gazebo_motor_model` for all the rest of the models only include the dynamics for the motor itself. For plane / multirotos there are separate plugins for the motor model and the fuselarge interaction with aerodynamics, and multiple motor elements are added depending on how many actuators there are in the system.

- This is more out of curiousity, as the thrust is modeled linear to the rotational velocity. Could you point us to any reason / literature that this was modeled this way? I am not an expert of UUVs, but I would expect that in a newtonian fluid (such as water) the thrust will increase quadratically compared to the rotational velocity of the srews. Does the linear model better fit in low thrust scenarios for the hippocampus?

Jaeyoung-Lim commented on models/uuv_hippocampus/uuv_hippocampus.sdf:
```diff
@@ -0,0 +1,645 @@
+  <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
+  <motorSpeedPubTopic>/motor_speed/0</motorSpeedPubTopic>
+  <rotorVelocitySlowdownSim>0.05</rotorVelocitySlowdownSim>
+</plugin>-->
```
Please remove the commented out parts

DanielDuecker replied:
those parts are now back in again and now spin the propellers/thruster

Jaeyoung-Lim commented on models/uuv_hippocampus/uuv_hippocampus.sdf:
```diff
@@ -0,0 +1,645 @@
+    <plugin name="gps_plugin" filename="libgazebo_gps_plugin.so">
+        <robotNamespace/>
+        <gpsNoise>true</gpsNoise>
+    </plugin>
```
Does the hippocampus have GPS?

DanielDuecker replied:
Our real world version does not have GPS since GPS does not work/is attenuated almost immediately underwater. Underwater localization is some kind of a night mare which I would prefer for the moment to keep out of the simulation
Hence we implemented "GPS" as a well integrated work around with fairly similar characteristics. However, this can be of course extended at some point but it will be quite a chunk of work to model the real world localization characteristics.

Jaeyoung-Lim replied:
Fair enough. I don't think we need to "fix" this, but still be aware the state estimation is using GPS data

DanielDuecker committed:
* remove the old hippocampus model
* updates on the new uuv_hippocampus model
	- add nicely spinning propeller
	- add red/green coloring to propellers -> port/starboard side

* Merge the new uuv_motor_model into the UUV_plugin, which now

  - computes the hydrodynamic effects (according to FOSSEN 2011 - Handbook of Marine Craft Hydrodynamics and Motion Control)
  - applies the Hydro effects  as forces/moments and are than applied to the vehicle
  - computes the forces/moments based on the motor commands

Jaeyoung-Lim reviewed (approved):
I am not really happy with the fact that the thrusters are included in the plugin and that you are not able to add more thrusters.

However, since this is specific to the hippocampus and @DanielDuecker is developing it, I think this is good to go on myside.
